session/session.go

```diff
@@ -167,18 +181,45 @@ func NewInactiveSession() *Session {
 	}
 }
 
-func (s *Session) Activate(ctx context.Context, i *identity.Identity, c lifespanProvider, authenticatedAt time.Time) error {
+func (s *Session) Activate(r *http.Request, i *identity.Identity, c lifespanProvider, authenticatedAt time.Time) error {
 	if i != nil && !i.IsActive() {
 		return ErrIdentityDisabled.WithDetail("identity_id", i.ID)
 	}
 
 	s.Active = true
-	s.ExpiresAt = authenticatedAt.Add(c.SessionLifespan(ctx))
+	s.ExpiresAt = authenticatedAt.Add(c.SessionLifespan(r.Context()))
 	s.AuthenticatedAt = authenticatedAt
 	s.IssuedAt = authenticatedAt
 	s.Identity = i
 	s.IdentityID = i.ID
 
+	agent := r.Header["User-Agent"]
```

Comment:
So this code sets the session information *at the time of session activation*, so when the session is usually issued.

The naming however does not indicate that, as it just says `client_ip_address`. In the original issue we discuss an idea for how to track a device over it's lifetime:

https://github.com/ory/kratos/issues/2091#issuecomment-1003414305

At the least I think these fields should be arrays! We don't have to track these values over their lifetime _just yet_ and could just add the IP during login, but the API should be including this use case which we might add in the future!

Reply:
Sounds good, I've tweaked the design a bit. We should be able to track list of the combo of UserAgent (client) - IP - Location. The model looks like below which is then embedded as a list in the Session.  

```go
type Metadata struct {
	ID uuid.UUID `json:"id" faker:"-" db:"id"`
	SessionID uuid.UUID `json:"-" faker:"-" db:"session_id"`
	IPAddress *string `json:"ip_address" db:"ip_address"`
	UserAgent *string `json:"user_agent" db:"user_agent"`
	Location *string `json:"location" db:"location"`
	CreatedAt time.Time `json:"seen_at" faker:"-" db:"created_at"`
        LastSeen time.Time `json:"last_seen" faker:"-" db:"last_seen"`
	NID uuid.UUID `json:"-"  faker:"-" db:"nid"`
}
```

```go
	// Metadata has history of all clients where the session was used
	Metadata []Metadata `json:"metadata" faker:"-" db:"-"`
```

The metadata records are inserted in a new table. The PR still needs work.

TODO - 

- [ ] `Upsert` should save metadata records
- [ ] `SaveSessionMetadata` should update the last seen only on metadata record that already exists for the same IP.

Reply:
Can you please address the following problems in the design:

1. Last seen is very hard to implement at scale, I would remove it for now
2. The session will get a metadata field which can be used to store arbitrary data (eg campaign source=facebook) Please rename the field to e.g device or agent or client
3. One user agent can have multiple IPs and locations, but one session usually has only one user agent (we should still allow multiple though)